ludwig/utils/remote_storage_options.py

```diff
@@ -0,0 +1,60 @@
+#! /usr/bin/env python
+# Copyright (c) 2021 Linux Foundation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import os
+from abc import ABCMeta
+from typing import Any, Dict
+
+from ludwig.constants import (
+    AWS_ACCESS_KEY_ID,
+    AWS_SECRET_ACCESS_KEY,
+    CLIENT_KWARGS,
+    ENDPOINT_URL,
+    KEY,
+    MLFLOW_S3_ENDPOINT_URL,
+    SECRET,
+)
+
+
+class BaseRemoteStorageOptions(metaclass=ABCMeta):
+    """Base Class for remote storage options to be used by fsspec Guidelines for storage options structure:
+
+    https://s3fs.readthedocs.io/en/latest/#s3-compatible-storage.
+    """
+
+    def __init__(self, key: str, secret: str, endpoint_url: str):
+        self.key = os.environ.get(key, None)
+        self.secret = os.environ.get(secret, None)
+        self.endpoint_url = os.environ.get(endpoint_url, None)
+
+    def get_key(self) -> str:
+        return self.key
+
+    def get_secret(self) -> str:
+        return self.secret
+
+    def get_client_kwargs(self) -> Dict[str, Any]:
+        return {ENDPOINT_URL: self.endpoint_url}
+
+    def get_storage_options(self):
+        return {KEY: self.get_key(), SECRET: self.get_secret(), CLIENT_KWARGS: self.get_client_kwargs()}
+
+
+class S3RemoteStorageOptions(BaseRemoteStorageOptions):
+    """Get credentials from environment variables."""
+
+    def __init__(self):
+        super().__init__(AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, MLFLOW_S3_ENDPOINT_URL)
```

Comment:
Doesn't make sense to me to have mlflow coupling here. Looks like the idea here is to pull storage options from certain environment variables when they're not passed in explicitly. However, this shouldn't be necessary, as fsspec already has a mechanism for plumbing credentials through the environment, either through the `use_credentials` helper that sets the `conf` dictionary, or through the `FSSPEC_` env vars:

https://filesystem-spec.readthedocs.io/en/latest/features.html#configuration

So I would rely on the built-in fsspec plumbing mechanisms instead of recreating it here.

Reply:
@tgaddair the `MLFLOW_S3_ENDPOINT_URL` is actually just a placeholder name and can be change to any other name. It doesn't actually end up involving any coupling with MLFLOW. I'll rename this environment variable to something more generic like S3_ENDPOINT_URL to make this more clear! 

I like the idea of using the FSSPEC_ env vars as a potential substitute to avoid plumbing S3 credentials in this particular case with endpoint URLs. Will update and clear this up so that we don't need to do any of this manually.

Reply:
Although following from the thread here (https://github.com/fsspec/s3fs/issues/432) where you've requested this before as well, it doesn't seem like there's a neat way to specify the endpoint URL since it needs to be set inside a nested dictionary. How do you feel about just using this as an environment variable for now but with a different name?